components/tidb_query/src/codec/mysql/time/datetime.rs

```diff
@@ -0,0 +1,1306 @@
+        let modulus = [
+            0,
+            12,
+            last_day_of_month(parts[0], parts[1]),
+            24,
+            60,
+            60,
+            1_000_000,
+        ];
+
+        for i in (1..7).rev() {
+            if parts[i] >= modulus[i] {
+                parts[i] -= modulus[i];
+                parts[i - 1] += 1;
+            }
+        }
+        Ok(())
+    }
+
+    pub fn parse(
+        ctx: &mut EvalContext,
+        input: &str,
+        time_type: TimeType,
+        fsp: u8,
+        round: bool,
+    ) -> Result<Time> {
+        let trimmed = input.trim();
+        if trimmed.is_empty() {
+            assert(false)?;
+        }
+
+        let components = split_components(trimmed)?;
+        match components.len() {
+            1 | 2 => {
+                let whole = parse_whole(components[0])?;
+
+                let (carry, frac) = if let Some(frac) = components.get(1) {
+                    // If we have a fractional part,
+                    // we expect the `whole` is in format: `yymmddhhmmss/yyyymmddhhmmss`.
+                    // Otherwise, the fractional part is meaningless.
+                    assert(components[0].len() == 12 || components[0].len() == 14)?;
+                    parse_frac(frac, fsp, round)?
+                } else {
+                    (false, 0)
+                };
+
+                let mut parts = concat_components(vec![&whole, &[frac]]);
+                if carry {
+                    assert(round_components(&mut parts).is_ok())?;
+                }
+
+                Time::from_slice(ctx, &parts, time_type, fsp)
+            }
+            3..=7 => {
+                let whole = std::cmp::min(components.len(), 6);
+                let mut parts: Vec<_> =
+                    components[..whole]
+                        .iter()
+                        .try_fold(vec![], |mut acc, part| -> Result<_> {
+                            acc.push(bytes_to_u32(part)?);
+                            Ok(acc)
+                        })?;
+                parts.resize(6, 0);
+
+                let (carry, frac) = if let Some(frac) = components.get(6) {
+                    parse_frac(frac, fsp, round)?
+                } else {
+                    (false, 0)
+                };
+                parts.push(frac);
+
+                if carry {
+                    assert(round_components(&mut parts).is_ok())?;
+                }
+                Time::from_slice(ctx, &parts, time_type, fsp)
+            }
+            _ => Err(Error::truncated()),
+        }
+    }
+}
+
+impl Time {
+    pub fn parse(
+        ctx: &mut EvalContext,
+        input: &str,
+        time_type: TimeType,
+        fsp: i8,
+        round: bool,
+    ) -> Result<Time> {
+        parser::parse(ctx, input, time_type, mysql::check_fsp(fsp)?, round)
+            .map_err(|_| Error::incorrect_datetime_value(input))
+    }
+}
+
+macro_rules! handle_zero_date {
+    ($datetime:expr, $ctx:expr) => {{
```

Comment:
Why not a function for this?

Reply:
If a zero `TimeArgs` acquired, we need to handle it with `ctx`.
If this session allows zero-date, we return it normally. Otherwise, we need to report errors or append some warnings into `ctx`.